topdown/graphql.go

```diff
@@ -0,0 +1,192 @@
+			return result, err
+		}
+	default:
+		return result, errGraphQLError(builtinName, "non-AST type fed into AST-to-JSON transformer.")
+	}
+	return result, nil
+}
+
+func builtinGraphQLParse(_ BuiltinContext, operands []*ast.Term, iter func(*ast.Term) error) error {
+	// Get the raw strings from each operand so that we can
+	// feed them to the GraphQL parser functions.
+	raw_query, err := builtins.StringOperand(operands[0].Value, 1)
+	if err != nil {
+		return err
+	}
+	raw_schema, err := builtins.StringOperand(operands[1].Value, 1)
+	if err != nil {
+		return err
+	}
+
+	// Generate ASTs/errors for the GraphQL schema and query.
+	schema, errQ := gqltop.LoadSchema(&gqlast.Source{Input: string(raw_schema)})
+	if errQ != nil {
+		return errQ.Unwrap()
+	}
+	query, errS := gqlparser.ParseQuery(&gqlast.Source{Input: string(raw_query)})
+	if errQ != nil {
+		return errS.Unwrap()
+	}
+
+	// Validate the query against the schema, erroring if there's an issue.
+	errs := gqlvalidator.Validate(schema, query)
+	if errs != nil {
+		return errGraphQLErrorList(ast.GraphQLParse.Name, errs)
+	}
+
+	// Transform the AST to JSON, then into an Object.
+	result, err := transformASTToTerm(ast.GraphQLParse.Name, query)
+	if err != nil {
+		return err
+	}
+
+	return iter(&result)
+}
+
+func builtinGraphQLParseQuery(_ BuiltinContext, operands []*ast.Term, iter func(*ast.Term) error) error {
+	raw, err := builtins.StringOperand(operands[0].Value, 1)
+	if err != nil {
+		return err
+	}
+
+	// Get the high-nested AST struct, along with any errors generated.
+	AST, errQ := gqlparser.ParseQuery(&gqlast.Source{Input: string(raw)})
+	if errQ != nil {
+		return errQ.Unwrap()
+	}
+
+	// Transform the AST to JSON, then into an Object.
+	result, err := transformASTToTerm(ast.GraphQLParse.Name, AST)
+	if err != nil {
+		return err
+	}
+
+	return iter(&result)
+}
+
+func builtinGraphQLParseSchema(_ BuiltinContext, operands []*ast.Term, iter func(*ast.Term) error) error {
+	raw, err := builtins.StringOperand(operands[0].Value, 1)
+	if err != nil {
+		return err
+	}
+
+	// Get the high-nested AST struct, along with any errors generated.
+	AST, errS := gqlparser.ParseSchema(&gqlast.Source{Input: string(raw)})
+	if errS != nil {
+		return errS.Unwrap()
+	}
+
+	// Transform the AST to JSON, then into an Object.
+	result, err := transformASTToTerm(ast.GraphQLParse.Name, AST)
+	if err != nil {
+		return err
+	}
+
+	return iter(&result)
+}
+
+func builtinGraphQLIsValid(_ BuiltinContext, operands []*ast.Term, iter func(*ast.Term) error) error {
+	// Get the raw strings from each operand so that we can
+	// feed them to the GraphQL parser functions.
+	raw_query, err := builtins.StringOperand(operands[0].Value, 1)
+	if err != nil {
+		return err
+	}
+	raw_schema, err := builtins.StringOperand(operands[1].Value, 1)
+	if err != nil {
+		return err
+	}
+
+	// Generate ASTs/errors for the GraphQL schema and query.
+	schema, errQ := gqltop.LoadSchema(&gqlast.Source{Input: string(raw_schema)})
```

Comment:
[nit] It's a style thing, but I think it's common to just overwrite `err` again and again. Let's do that here, too: 
```suggestion
	schema, err := gqltop.LoadSchema(&gqlast.Source{Input: string(raw_schema)})
```

Reply:
I changed to a different error variable name there because `gqltop.LoadSchema` returns a `gqlerror.Error` struct (not a normal `error`!), and this resulted in the staticcheck linter suggesting the "is nil?" condition following would always be true. Exact warning from staticcheck quoted below:

> this comparison is always true (SA4023)
	graphql.go:168:2: the lhs of the comparison gets its value from here and has a concrete type

I worked around the linter warning by using a name that had not been assigned yet (and thus would be inferred to have the richer `gqlerror.Error` type). If the value is non-nil, then the code plucks out the underlying `error` using the `gqlerror.Error` struct's `.Unwrap()` method.

Is there a better way to work around this kind of thing, other than a different variable name?

Reply:
Hah that's interesting. Quite a pitfall, https://staticcheck.io/docs/checks#SA4023.

The library's `*gqlerror.Error` error type satisfies the `error` interface: [`Error() string`](https://pkg.go.dev/github.com/vektah/gqlparser/v2@v2.4.2/gqlerror#Error.Error)... so how about
```go
// loadSchema is a helper function wrapping gqltop.LoadSchema
func loadSchema(schema string) (*validator.Schema, error) {
	schema, err := gqltop.LoadSchema(&gqlast.Source{Input: schema})
	if err != nil { // struct pointer!
		return nil, err // now it's converted to the error interface
	}
	return schema, nil // that's an OK err==nil thing
}
```

That helper could then be used as `schema, err := loadSchema(...); if err != nil { ... }`, I think -- haven't tried it, sorry 😅

Reply:
@srenatus Thank you for this suggestion! I'll look into implementing something like this shortly. :smile:

Reply:
Implemented this approach for all instances of `gqlerror.Error` and `gqlerror.List` (an error list type) that come from the GraphQL parser. The code's a bit cleaner as a result.